### PR TITLE
Remove deprecated items as scheduled in preparation of v1.12 (except `concatenatePointCloud`)

### PR DESCRIPTION
**Pull request by larshg**

### CODE REVIEW

File: common/include/pcl/point_cloud.h [hunk 1]
```diff
@@ -703,26 +695,10 @@ namespace pcl
       inline Ptr
       makeShared () const { return Ptr (new PointCloud<PointT> (*this)); }
 
-    protected:
-      /** \brief This is motivated by ROS integration. Users should not need to access mapping_.
-        * \todo Once mapping_ is removed, erase the explicitly defined copy constructor in PointCloud.
```
Comment (larshg) [inline common/include/pcl/point_cloud.h:708]: I think I overlooked this.

File: segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h [hunk 2]
```diff
@@ -194,71 +194,25 @@ namespace pcl
...
-
-      using Ptr = shared_ptr<EuclideanClusterComparator<PointT, PointNT, PointLT> >;
-      using ConstPtr = shared_ptr<const EuclideanClusterComparator<PointT, PointNT, PointLT> >;
+      using Ptr = shared_ptr<EuclideanClusterComparator<PointT, PointLT> >;
+      using ConstPtr = shared_ptr<const EuclideanClusterComparator<PointT, PointLT> >;
 
       using experimental::EuclideanClusterComparator<PointT, PointLT>::setExcludeLabels;
 
       /** \brief Default constructor for EuclideanClusterComparator. */
-      PCL_DEPRECATED(1, 12, "remove PointNT from template parameters")
```
Comment (larshg) [inline segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:216]: Is what I have done, the intended thing to do?

**Comment by kunaltyagi** [comment]
Should we make this dependent on #4336?

**Comment by larshg** [comment]
Yes, to test the deprecated macro first?

File: segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h (hunk 2 above)
Comment (kunaltyagi) [reply segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:216]: At a glance, yes

File: test/common/test_io.cpp [hunk 3]
```diff
@@ -197,57 +190,14 @@ TEST (PCL, concatenatePointCloud)
...
-  for (int i = 0; i < int (cloud_xyz_rgb.size ()); ++i)
-  {
-    EXPECT_XYZ_EQ (cloud_all[cloud_xyz_rgba.size () + i], cloud_xyz_rgb[i]);
-    EXPECT_FLOAT_EQ (cloud_all[cloud_xyz_rgba.size () + i].r, 0);
-    EXPECT_FLOAT_EQ (cloud_all[cloud_xyz_rgba.size () + i].g, 0);
-    EXPECT_FLOAT_EQ (cloud_all[cloud_xyz_rgba.size () + i].b, 0);
-    EXPECT_EQ (cloud_all[cloud_xyz_rgba.size () + i].rgba, 0);
-  }
-
   // _ vs _
```
[the remark this answers is not in the document]
Comment (kunaltyagi) [reply test/common/test_io.cpp:193]: I'm in favor of a test where we concatenate 2 point clouds: 1 over PointXYZ and one over XYZRGBA with RGBA skipped. If my proposed test is present, remove this test else modify

[the remark this answers is not in the document]
Comment (larshg) [reply test/common/test_io.cpp:193]: Well, the method doesn't support skipping a field?

It just compares all fields are equal and in the same order, then it copies. Else it reject it if there are any differences, or a "_" field is encountered.

[the remark this answers is not in the document]
Comment (kunaltyagi) [reply test/common/test_io.cpp]: It does. It skips "_" in lines 270 and 287

[the remark this answers is not in the document]
Comment (larshg) [reply test/common/test_io.cpp]: The concatenate it self doesn't skip fields when it hit the first "_" in either pointcloud it skips checking the last field of the second pointcloud and hence, it find a missmatch in number of fields and just print this `PCL_ERROR ("[pcl::PCLPointCloud2::concatenate] Number of fields to copy in cloud1 (%u) != Number of fields to copy in cloud2 (%u)\n", i, j); ` and returns false:

https://github.com/PointCloudLibrary/pcl/blob/146ad42cdabaff2b207bd9a3399b8ba0d0b29e7c/common/src/PCLPointCloud2.cpp#L105-L136

[the remark this answers is not in the document]
Comment (larshg) [reply test/common/test_io.cpp]: And if one have a PoinXYZI pointcloud and PointXYZRGBA pointcloud and set both last field names to "_" it would end up doing insert with one kind of data, into a cloud with other data, since it would conclude it is `simplelayout`:
https://github.com/PointCloudLibrary/pcl/blob/146ad42cdabaff2b207bd9a3399b8ba0d0b29e7c/common/src/PCLPointCloud2.cpp#L83-L88

And this skips the entire checking of "_" values, continuing with just copying the data:

https://github.com/PointCloudLibrary/pcl/blob/146ad42cdabaff2b207bd9a3399b8ba0d0b29e7c/common/src/PCLPointCloud2.cpp#L145-L149

File: common/include/pcl/point_cloud.h [hunk 4]
```diff
@@ -703,26 +695,10 @@ namespace pcl
...
-        */
-      PCL_DEPRECATED(1, 12, "rewrite your code to avoid using this protected field") shared_ptr<MsgFieldMap> mapping_;
-
-      friend shared_ptr<MsgFieldMap>& detail::getMapping<PointT>(pcl::PointCloud<PointT> &p);
-
     public:
```
Comment (SergioRAgostinho) [inline common/include/pcl/point_cloud.h:698]: delete also this line

File: common/include/pcl/point_cloud.h (hunk 1 above)
Comment (SergioRAgostinho) [reply common/include/pcl/point_cloud.h:708]: Yup. Please remove also 
https://github.com/PointCloudLibrary/pcl/blob/29c6f61454c6a279c0df8c3efd1334be19446293/common/include/pcl/point_cloud.h#L181-L193

File: gpu/kinfu/tools/kinfu_app.cpp [hunk 5]
```diff
@@ -132,14 +132,14 @@ namespace pcl
           * \return true if the operation was successful (the handler is capable and 
           * the input cloud was given as a valid pointer), false otherwise
           */
```
Comment (SergioRAgostinho) [inline gpu/kinfu/tools/kinfu_app.cpp:134]: You need to update the docstring. Any idea why this change was required in this PR? I would rather do this in a separate PR. Otherwise it's gonna get buried here.

File: gpu/kinfu_large_scale/tools/color_handler.h [hunk 6]
```diff
@@ -63,14 +63,13 @@ namespace pcl
         capable_ = true;
       }
             
-      bool 
-      getColor (vtkSmartPointer<vtkDataArray> &scalars) const override
+      vtkSmartPointer<vtkDataArray>
+      getColor () const override
```
Comment (SergioRAgostinho) [inline gpu/kinfu_large_scale/tools/color_handler.h:67]: move this change set to a new PR.

File: segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h (hunk 2 above)
Comment (SergioRAgostinho) [reply segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:216]: Not fully. You should delete the current `EuclideanClusterComparator` class entirely and move `experimental::EuclideanClusterComparator<PointT, PointLT>` ->  `EuclideanClusterComparator<PointT, PointLT>`

File: segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h [hunk 7]
```diff
@@ -194,79 +194,29 @@ namespace pcl
...
-
       float angular_threshold_;
   };
 
-  template<typename PointT, typename PointLT>
-  class EuclideanClusterComparator<PointT, PointLT, deprecated::T>
-    : public experimental::EuclideanClusterComparator<PointT, PointLT> {};
+  template<typename PointT>
+  class EuclideanClusterComparator<PointT>
+    : public experimental::EuclideanClusterComparator<PointT> {};
```
Comment (SergioRAgostinho) [inline segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:221]: To delete entirely.

File: test/common/test_io.cpp [hunk 8]
```diff
@@ -129,15 +129,35 @@ TEST (PCL, copyPointCloud)
...
+  cloud_xyz2.push_back(pt_xyz);
+  cloud_xyz2.push_back(pt_xyz);
+  cloud_xyz2.push_back(pt_xyz);
+  cloud_xyz2.push_back(pt_xyz);
+
+  CloudXYZ cloud_xyz_out;
+  pcl::concatenate<pcl::PointXYZ>(cloud_xyz1, cloud_xyz2, cloud_xyz_out);
+
+  EXPECT_EQ(cloud_xyz_out.size(), cloud_xyz1.size() + cloud_xyz2.size());
+  for (int i = 0; i < int(cloud_xyz1.size()); ++i)
```
Comment (SergioRAgostinho) [inline test/common/test_io.cpp:151]: use index_t and use a static cast please.

File: test/common/test_io.cpp [hunk 9]
```diff
@@ -129,15 +129,35 @@ TEST (PCL, copyPointCloud)
...
+  {
+    EXPECT_XYZ_EQ(cloud_xyz_out[i], cloud_xyz1[i]);
+  }
+  for (int i = 0; i < int(cloud_xyz2.size()); ++i)
```
Comment (SergioRAgostinho) [inline test/common/test_io.cpp:155]: same here.

File: test/common/test_io.cpp [hunk 10]
```diff
@@ -149,181 +169,24 @@ TEST (PCL, concatenatePointCloud)
...
   cloud_xyz_rgba2.push_back (pt_xyz_rgba2);
 
-  pcl::PCLPointCloud2 cloud1, cloud2, cloud_out, cloud_out2, cloud_out3, cloud_out4;
-  pcl::toPCLPointCloud2 (cloud_xyz_rgba, cloud1);
-  pcl::toPCLPointCloud2 (cloud_xyz_rgba2, cloud2);
-
+  CloudXYZRGBA cloud_out;
   // Regular
-  pcl::concatenatePointCloud (cloud1, cloud2, cloud_out);
+  pcl::concatenate<pcl::PointXYZRGBA> (cloud_xyz_rgba, cloud_xyz_rgba2, cloud_out);
```
Comment (SergioRAgostinho) [inline test/common/test_io.cpp:174]: did you need to explicitly specify the template type?

File: test/common/test_io.cpp [hunk 11]
```diff
@@ -149,181 +169,24 @@ TEST (PCL, concatenatePointCloud)
...
-  pcl::fromPCLPointCloud2 (cloud_out4, cloud_all);
-
-  EXPECT_EQ (cloud_all.size (), cloud_xyz_rgba.size () + cloud_xyz_rgba2.size ());
-  for (int i = 0; i < int (cloud_xyz_rgb.size ()); ++i)
-  {
-    EXPECT_XYZ_EQ (cloud_all[i], cloud_xyz_rgb[i]);
-    // Data doesn't get modified
-    EXPECT_RGB_EQ (cloud_all[i], cloud_xyz_rgb[i]);
-    EXPECT_EQ (cloud_all[i].rgba, cloud_xyz_rgb[i].rgba);
-  }
```
Comment (SergioRAgostinho) [inline test/common/test_io.cpp:313]: What happened here?

File: test/common/test_io.cpp [hunk 12]
```diff
@@ -419,19 +282,6 @@ TEST (PCL, concatenatePointCloud2)
...
-
-  cloud1.fields[rgb_idx].name = "_";
-  pcl::toPCLPointCloud2 (cloud_xyz_rgb, cloud2);
-  EXPECT_FALSE (pcl::concatenate (cloud2, cloud1, cloud_out3));
-  cloud1.fields[rgb_idx].name = "rgba";
-
-  // regular vs _
-  rgb_idx = pcl::getFieldIndex (cloud2, "rgb");
-  cloud2.fields[rgb_idx].name = "_";
-  EXPECT_FALSE (pcl::concatenate (cloud2, cloud1, cloud_out4));
```
Comment (SergioRAgostinho) [inline test/common/test_io.cpp:434]: Also what happened here?

**Review by SergioRAgostinho — CHANGES_REQUESTED** [review]

File: test/common/test_io.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (SergioRAgostinho) [reply test/common/test_io.cpp]: Can we move this discussion to a separate PR and revert all changes with respect to it there. The issue doesn't seem superficial.

File: gpu/kinfu/tools/kinfu_app.cpp (hunk 5 above)
Comment (larshg) [reply gpu/kinfu/tools/kinfu_app.cpp:134]: It's currently overriding a deprecated method https://github.com/PointCloudLibrary/pcl/blob/5e8f8beb4f6793b507154b80fa269f02eb51f41c/visualization/include/pcl/visualization/point_cloud_color_handlers.h#L113-L122

But that doesn't seem to spawn any warnings.

I'll move it to a separate PR.

**Comment by larshg** [comment]
Reverted kinfu and concatenatePointCloud.

Hopefully the old EuclideanClusterComparator is correctly removed now.

**Comment by larshg** [comment]
Need to rebase this on master now, to avoid the getColor error in kinfu.

https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=17345&view=logs&jobId=d0a7bda2-01ee-550e-f56c-ab8664f27120&j=d0a7bda2-01ee-550e-f56c-ab8664f27120&t=e76c7bc5-5686-5622-3ba1-e36139d9a40b

**Comment by kunaltyagi** [comment]
Rebase now or merge now and rebase later: both are ok

**Review by kunaltyagi — COMMENTED** [review]
26/27

Should we increase the PCL version here?

**Comment by larshg** [comment]
If we increase, we will get errors instead of warnings right? And there are still some deprecations about the concatenatePointCloud which are not handled.

Which would cripple the CI?

**Comment by kunaltyagi** [comment]
Got it. In that case, let's not 😄

**Review by mvieth — APPROVED** [review]

File: segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h [hunk 13]
```diff
@@ -48,225 +48,141 @@
...
-  } // namespace experimental
-
-
-  /** \brief EuclideanClusterComparator is a comparator used for finding clusters based on euclidian distance.
-    *
-    * \author Alex Trevor
-    */
-  template<typename PointT, typename PointNT, typename PointLT = deprecated::T>
-  class EuclideanClusterComparator : public experimental::EuclideanClusterComparator<PointT, PointLT>
+  template<typename PointT, typename PointLT = pcl::Label>
```
Comment (kunaltyagi) [inline segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:51]: missing docstring

File: segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h [hunk 14]
```diff
@@ -48,225 +48,141 @@
...
+      using ExcludeLabelSetConstPtr = shared_ptr<const ExcludeLabelSet>;
 
       /** \brief Default constructor for EuclideanClusterComparator. */
-      PCL_DEPRECATED(1, 12, "remove PointNT from template parameters")
       EuclideanClusterComparator ()
-        : normals_ ()
-        , angular_threshold_ (0.0f)
+        : distance_threshold_ (0.005f)
+        , depth_dependent_ ()
       {}
```
Comment (kunaltyagi) [inline segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:77]: Replace with `EuclideanClusterComparator() = default;` and use default init for the variables?

File: segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h [hunk 15]
```diff
@@ -48,225 +48,141 @@
...
+        distance_threshold_ = distance_threshold;
+        depth_dependent_ = depth_dependent;
       }
 
-      /** \brief Get the angular threshold in radians for difference in normal direction between neighboring points, to be considered part of the same plane. */
-      PCL_DEPRECATED(1, 12, "EuclideadClusterComparator never actually used normals and angular threshold, this function has no effect on the behavior of the comparator. It is deprecated and will be removed in future releases.")
+      /** \brief Get the distance threshold in meters (d component of plane equation) between neighboring points, to be considered part of the same plane. */
       inline float
-      getAngularThreshold () const { return (std::acos (angular_threshold_) ); }
+      getDistanceThreshold () const
```
Comment (kunaltyagi) [inline segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:100]: getter for depth dependent missing

File: segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h [hunk 16]
```diff
@@ -48,225 +48,141 @@
...
+      {
+        return (distance_threshold_);
+      }
+
+      /** \brief Set label cloud
+        * \param[in] labels The label cloud
+        */
+      void
+      setLabels (const PointCloudLPtr& labels)
```
Comment (kunaltyagi) [inline segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:109]: getter for labels_ missing

File: segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h [hunk 17]
```diff
@@ -179,7 +195,7 @@ namespace pcl
         */
       ExcludeLabelSetConstPtr exclude_labels_;
 
-      float distance_threshold_;
+      float distance_threshold_ = 0.005f;
 
       bool depth_dependent_;
```
Comment (kunaltyagi) [inline segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:200]: true or false? I know it didn't have a default before, but can we add one (false would be like 0-init)? It'll be less headaches in future

Comment (larshg) [reply segmentation/include/pcl/segmentation/euclidean_cluster_comparator.h:200]: I just anticipated it would be false, but I'll add it explicitly 👍

**Review by kunaltyagi — APPROVED** [review]

**Comment by kunaltyagi** [comment]
Today's a green day for @larshg

**Comment by larshg** [comment]
I have squashed correction commits to their respective commits. Please merge either of you, @kunaltyagi or @mvieth. So we can get rid of a lot of warnings on windows at least 😄
